Build CairoSVG 1.0.22, needed for python<3

MutantPlatypus opened the pull request:
Per the CairoSVG change log, support for Python 2 was dropped at cairosvg==2.0.0.  The last version that supported `python<3` was 1.0.22.

This recipe also includes all optional dependencies.

Fixes Issue #22

<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
1.0.22 has three optional dependencies and two required.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

carlodri commented on recipe/meta.yaml:
```diff
@@ -1,33 +1,33 @@
 requirements:
   host:
-    - python >=3.5
+    - python <3
```
is this constraint really required? or could it be simply `python`? (same below in `run` section)

MutantPlatypus replied:
Hmm, [setup.py at the 1.0.22 tag](https://github.com/Kozea/CairoSVG/blob/1.0.22/setup.py) only declares support up to Python 3.4.  Could be risky to try it on `python>3.4`, and the whole point of bothering to take this back to 1.0.22 was to make sure it was to offer a solution for `python<3`.  I see now that this might also be a solution for `python<=3.4`, since the latest versions are pinned to `python>=3.5`.  I'll try changing this to `python<=3.4` and see how it goes, but  I want to make sure the CI services also test on python<3

MutantPlatypus replied:
I think I get it now. `cairosvg` is pure python, but its dependencies are not. I should let the package manager figure the dependencies out: We just need to tell conda-forge to build and keep 1.0.22. That way, the dependency solver will find it for `python<3`. If there are problems in the dependencies not being available for some OSes for `python<3`, the dependency solver will notice. Issues will be reported to this or the dependency feedstocks and they can be fixed in the feedstocks of the respective dependencies.

This still allows a user to try and use `cairosvg==1.0.22` in a python 3 environment if they see a need to, since it's not explicitly forbidden by `cairosvg`.

That still leaves the question of the optional dependencies. I think we should include them when a user asks conda to install `cairosvg`: It is most likely to work when all of the optional deps are there. On top of that, these optional deps turned into mandatory ones in later releases, so the maintainers of `cairosvg` thought the features were important enough to include. It also avoids the problem of the user needing to figure out that, even though on python 3 they just have to `conda install cairosvg`, on python 2 they need to `conda install cairosvg lxml cssselect tinycss` because their code uses some feature of `cairosvg` that required the optional deps. What do you think? ~~I'll force push a clean commit history to PR#23 and reopen that PR so these experiments remain available.~~ I can't reopen a pull request after force pushing, so I'll pop it here and open it for review.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * `noarch` packages can't have selectors. If the selectors are necessary, please remove `noarch: python`.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

MutantPlatypus commented:
I think I get it now. `cairosvg` is pure python, but its dependencies are not. I should let the package manager figure the dependencies out: We just need to tell conda-forge to build and keep 1.0.22. That way, the dependency solver will find it for `python<3`. If there are problems in the dependencies not being available for some OSes for `python<3`, the dependency solver will notice. Issues will be reported to this or the dependency feedstocks and they can be fixed in the feedstocks of the respective dependencies.

This still allows a user to try and use `cairosvg==1.0.22` in a python 3 environment if they see a need to, since it's not explicitly forbidden by `cairosvg`.

That still leaves the question of the optional dependencies. I think we should include them when a user asks conda to install `cairosvg`: It is most likely to work when all of the optional deps are there. On top of that, these optional deps turned into mandatory ones in later releases, so the maintainers of `cairosvg` thought the features were important enough to include. It also avoids the problem of the user needing to figure out that, even though on python 3 they just have to `conda install cairosvg`, on python 2 they need to `conda install cairosvg lxml cssselect tinycss` because their code uses some feature of `cairosvg` that required the optional deps. What do you think? ~~I'll force push a clean commit history to PR#23 and reopen that PR so these experiments remain available.~~ I can't reopen a pull request after force pushing, so I'll pop it here and open it for review.

MutantPlatypus committed:
Build CairoSVG 1.0.22, needed for python<3

Per the CairoSVG change log, support for Python 2 was dropped at
cairosvg==2.0.0.  The last version that supported `python<3` was
1.0.22.

This recipe also includes all optional dependencies.

MutantPlatypus commented:
@conda-forge-admin, please rerender

conda-forge-admin committed:
MNT: Re-rendered with conda-build 3.18.9, conda-smithy 3.4.8, and conda-forge-pinning 2019.09.27

carlodri commented:
@MutantPlatypus as far as I understand you are correct in all your observations. The question of optional dependencies is summarized here https://github.com/conda/conda/issues/7502. At the moment, it is up to the maintainers to decide and I had decided to include all of them as default to avoid user headaches.

~Why did you remove the optional deps here? Sorry if I missed something.~ Sorry, I indeed missed something! Yes, I would leave the dependencies inside the package as you did!

MutantPlatypus commented:
In that case, I have no further changes I would propose for this CR, it is ready for merging.  In fact, it doesn't help to wait for #17:  1.0.22 did not pass it's own testing in its repo.  I wouldn't expect many of the tests being proposed as a solution to that to pass, either.

MutantPlatypus commented:
@conda-forge-admin, please rerender

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to re-render for you, but it looks like there was nothing to do.

MutantPlatypus commented:
Hi @carlodri, thanks for your support.  I learned a lot trying to get this going.  I have a couple of other discussions going so it might be a little confusing, but this PR is ready for review separate from other issues and discussions.  I even ran this build in both WSL and Windows 10 x86_64 and all SVGs render successfully.

If I have time I would like to change the tests from shell commands to a Python script so it's easier to run on other OSes, then I would like to find out exactly which dependency is causing just one test to fail if I don't pin this to `python<3`.  (It also takes _forever_ for the dependency solver to finish if it's not pinned to py2k.) And also get both 1.0.22 and main testing on all OSes.  But I can do that in later PRs.

MutantPlatypus committed:
Verify that all SVG files in test folder rasterize

MutantPlatypus committed:
Pin to py2k

carlodri commented:
Hi @MutantPlatypus, is this PR still relevant to you? Or can we close it?

carlodri commented:
Closing due to inactivity, please reopen if relevant.
